src/otg/preprocess/finngen/study_index.py

```diff
@@ -0,0 +1,128 @@
+from urllib.request import urlopen
+
+import pyspark.sql.functions as f
+from pyspark.sql import SparkSession
+
+from otg.common.schemas import parse_spark_schema
+from otg.dataset.study_index import StudyIndex
+
+if TYPE_CHECKING:
+    from pyspark.sql import DataFrame
+    from pyspark.sql.types import StructType
+
+logging.basicConfig(level=logging.INFO, format="%(asctime)s %(message)s")
+
+
+@dataclass
+class FinnGenStudyIndex(StudyIndex):
+    """Study index dataset from FinnGen.
+
+    The following information is aggregated/extracted:
+
+    - Study ID in the special format (FINNGEN_R9_*)
+    - Trait name (for example, Amoebiasis)
+    - Number of cases and controls
+    - Link to the summary statistics location
+
+    Some fields are also populated as constants, such as study type and the initial sample size.
+    """
+
+    @classmethod
+    def get_schema(cls: type[FinnGenStudyIndex]) -> StructType:
+        """Provides the schema for the FinnGenStudyIndex dataset."""
+        return parse_spark_schema("study_index.json")
+
+    @classmethod
+    def from_source(
+        cls: type[FinnGenStudyIndex],
+        finngen_studies: DataFrame,
+        finngen_release_prefix: str,
+        finngen_summary_stats_url_prefix: str,
+        finngen_summary_stats_url_suffix: str,
+    ) -> FinnGenStudyIndex:
+        """This function ingests study level metadata from FinnGen.
+
+        Args:
+            finngen_studies (DataFrame): FinnGen raw study table
+            finngen_release_prefix (str): Release prefix pattern.
+            finngen_summary_stats_df_url_prefix (str): URL prefix for summary statistics location.
+            finngen_summary_stats_df_url_suffix (str): URL prefix suffix for summary statistics location.
+
+        Returns:
+            FinnGenStudyIndex: Parsed and annotated FinnGen study table.
+        """
+        return cls(
+            _df=finngen_studies.select(
+                f.concat(f.lit(f"{finngen_release_prefix}_"), f.col("phenocode")).alias(
+                    "studyId"
+                ),
+                f.col("phenostring").alias("traitFromSource"),
+                f.col("num_cases").alias("nCases"),
+                f.col("num_controls").alias("nControls"),
+                (f.col("num_cases") + f.col("num_controls")).alias("nSamples"),
+                f.lit(finngen_release_prefix).alias("projectId"),
+                f.lit("gwas").alias("studyType"),
+                f.lit(True).alias("hasSumstats"),
+                f.lit("377,277 (210,870 females and 166,407 males)").alias(
+                    "initialSampleSize"
+                ),
+                f.concat(
+                    f.lit(finngen_summary_stats_url_prefix),
+                    f.col("phenocode"),
+                    f.lit(finngen_summary_stats_url_suffix),
+                ).alias("summarystatsLocation"),
+            ),
+            _schema=cls.get_schema(),
+        )
+
+
+def ingest_finngen_study_index(
```

Comment:
we are placing this logic in (e.g) `src/otg/finngen.py`. They are the definition of the "steps" and usually contain the minimum number of function calls to get the ETL functionality: read datasets, transform them, write the outputs. This separation of concerns helps to decouple the data model from the actual execution. For example, we pretty much don't have a spark session in any `otg/dataset`, `otg/method` or `otg/datasource`. Within these directories we should be aiming for pretty much 100% unit testing coverage, but the steps represent integration pipelines which are much less likely to be unit tested.

Reply:
Agreed. The layout of this PR is now in line with existing architecture, so fixed.